### PR TITLE
recipes: add election and double barrier recipes

### PR DESCRIPTION
**Pull request by heyitsanthony**
rely on leases so these weren't included in the last batch

### CODE REVIEW

File: contrib/recipes/double_barrier.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/storage/storagepb"
+)
+
+// DoubleBarrier blocks processes on Enter until an expected count enters, then
+// blocks again on Leave until all processes have left.
+type DoubleBarrier struct {
+	client *EtcdClient
+	key    string //key for the collective barrier
```
Comment (xiang90) [inline contrib/recipes/double_barrier.go]: space between "//" and "key"

File: contrib/recipes/double_barrier.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	b.myKey = ek
+
+	resp, err := NewRange(b.client, b.key+"/waiters").Prefix()
+	if err != nil {
+		return err
+	}
+
+	if len(resp.Kvs) > b.count {
+		log.Panicf("count is %v, expected no more than %v", len(resp.Kvs), b.count)
```
Comment (xiang90) [inline contrib/recipes/double_barrier.go]: return error instead of panic?

File: contrib/recipes/election.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// Leader returns the leader value for the current election.
+func (e *Election) Leader() (string, error) {
+	resp, err := NewRange(e.client, e.keyPrefix).FirstCreate()
+	if err != nil {
+		return "", err
+	} else if len(resp.Kvs) == 0 {
+		// no leader currently elected
+		return "", nil
```
Comment (xiang90) [inline contrib/recipes/election.go]: we should return no leader error here?

File: contrib/recipes/lease.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/lease"
+)
+
+// only keep one ephemeral lease per connection
+var connLeases connLeaseMgr = connLeaseMgr{leases: make(map[*grpc.ClientConn]*leaseKeepAlive)}
```
Comment (xiang90) [inline contrib/recipes/lease.go]: i feel the abstraction should not be at grpc.ClientConn layer. But it is fine for this pr.

File: contrib/recipes/lease.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+func SessionLease(conn *grpc.ClientConn) (lease.LeaseID, error) {
+	return connLeases.sessionLease(conn, 120)
+}
+
+func SessionLeaseTTL(conn *grpc.ClientConn, ttl int64) (lease.LeaseID, error) {
+	return connLeases.sessionLease(conn, ttl)
+}
+
+// StopSessionLease ends the refresh for the session lease.
+func StopSessionLease(conn *grpc.ClientConn) {
```
Comment (xiang90) [inline contrib/recipes/lease.go]: when do we prefer this to Revoke?

Comment (xiang90) [inline contrib/recipes/lease.go]: Transfer the ownership of leases?

File: contrib/recipes/lease.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	connLeases.mu.Unlock()
+	if lka != nil {
+		lka.donec <- struct{}{}
+		<-lka.donec
+	}
+}
+
+// RevokeSessionLease revokes the session lease.
+func RevokeSessionLease(conn *grpc.ClientConn) (err error) {
```
Comment (xiang90) [inline contrib/recipes/lease.go]: I feel we should have the ability to revoke a specific lease by a given leaseID.

Comment (heyitsanthony) [inline contrib/recipes/lease.go]: Sure. Defer until something needs it, though?

Comment (xiang90) [inline contrib/recipes/lease.go]: sure.

**Comment by heyitsanthony** [comment]
OK, all comments addressed. /cc @xiang90

**Comment by xiang90** [comment]
LGTM. Thanks!

**Comment by xiang90** [comment]
squash the commits :)

**Comment by heyitsanthony** [comment]
all squashed, merging on greenlight

**Pull request merged**